packages/datagateway-download/src/downloadApiHooks.ts

```diff
@@ -250,3 +327,434 @@ export const useDatafileCounts = (
+        }
+
+        return () =>
+          queryClient.setQueryData(QueryKey.DOWNLOADS, prevDownloads);
+      },
+
+      onSuccess: async (_, { downloadId, deleted }) => {
+        if (!deleted) {
+          // download is restored (un-deleted), fetch the download info
+          const restoredDownload = await getDownload(downloadId, {
+            facilityName: downloadSettings.facilityName,
+            downloadApiUrl: downloadSettings.downloadApiUrl,
+          });
+
+          if (restoredDownload) {
+            queryClient.setQueryData<Download[] | undefined>(
+              QueryKey.DOWNLOADS,
+              (downloads) => downloads && [...downloads, restoredDownload]
+            );
+          }
+        }
+      },
+
+      onError: (error, _, rollback) => {
+        handleICATError(error);
+        if (rollback) rollback();
+      },
+
+      retry: (failureCount, error) => {
+        // if we get 431 we know this is an intermittent error so retry
+        return error.code === '431' && failureCount < 3;
+      },
+    }
+  );
+};
+
+export const useDownloadTypeStatuses = <TData = DownloadTypeStatus>({
+  downloadTypes,
+  ...queryOptions
+}: {
+  downloadTypes: string[];
+} & UseQueryOptions<DownloadTypeStatus, AxiosError, TData>): UseQueryResult<
+  TData,
+  AxiosError
+>[] => {
+  // Load the download settings for use
+  const downloadSettings = React.useContext(DownloadSettingsContext);
+  const [t] = useTranslation();
+
+  const queryCount = downloadTypes.length;
+  const loadedQueriesCount = React.useRef(0);
+  const downloadTypesWithError = React.useRef<string[]>([]);
+
+  function broadcastError(message: string): void {
+    document.dispatchEvent(
+      new CustomEvent(MicroFrontendId, {
+        detail: {
+          type: NotificationType,
+          payload: {
+            severity: 'error',
+            message,
+          },
+        },
+      })
+    );
+  }
+
+  function handleQueryError(downloadType: string): void {
+    downloadTypesWithError.current.push(downloadType);
+
+    if (loadedQueriesCount.current === queryCount) {
+      if (downloadTypesWithError.current.length === queryCount) {
+        broadcastError(t('downloadConfirmDialog.access_methods_error'));
+      } else {
+        downloadTypesWithError.current.forEach((type) => {
+          broadcastError(
+            t('downloadConfirmDialog.access_method_error', {
+              method: type.toUpperCase(),
+            })
+          );
+        });
+      }
+    }
+  }
+
+  const queryConfigs: UseQueryOptions<
+    DownloadTypeStatus,
+    AxiosError,
+    TData
+  >[] = downloadTypes.map((type) => ({
+    queryKey: [QueryKey.DOWNLOAD_TYPE_STATUS, type],
+    queryFn: () =>
+      getDownloadTypeStatus(type, {
+        facilityName: downloadSettings.facilityName,
+        downloadApiUrl: downloadSettings.downloadApiUrl,
+      }),
+    onSettled: (_, error) => {
+      loadedQueriesCount.current += 1;
+      if (error) handleQueryError(type);
```

Comment:
I believe moving the error handling code into the `onSettled` call changes the original behaviour - before it waited until all the queries returned, and then either sent the `downloadConfirmDialog.access_methods_error` or sent individual `downloadConfirmDialog.access_method_error`. Now, we'll end up sending `downloadConfirmDialog.access_method_error` when each request fails, and then finally send `downloadConfirmDialog.access_methods_error` once they've all returned in an error state. This is potentially problematic as it could spam the user with lots of error messages if they're all unable to be fetched. I wonder if this is therefore better handled in a `useEffect` in `downloadConfirmDialog`...

Reply:
The `handleQueryError` waits until `loadedQueriesCount` matches the query count to either send individual errors or broadcast the `'downloadConfirmDialog.access_methods_error'`:

```js
if (loadedQueriesCount.current === queryCount) {
}
```

Reply:
Ah, I missed this so that's good :)